libkbfs/conflict_resolver.go

```diff
@@ -3070,6 +3013,114 @@ func (cr *ConflictResolver) calculateResolutionUsage(ctx context.Context,
 		md.SetDiskUsage(md.DiskUsage() - size)
 	}
 
+	return nil
+}
+
+// addUnrefToFinalResOp makes a resolutionOp at the end of opsList if
+// one doesn't exist yet, and then adds the given pointer as an unref
+// block to it.
+func addUnrefToFinalResOp(ops opsList, ptr BlockPointer) opsList {
+	resOp, ok := ops[len(ops)-1].(*resolutionOp)
+	if !ok {
+		resOp = newResolutionOp()
+		ops = append(ops, resOp)
+	}
+	resOp.AddUnrefBlock(ptr)
+	return ops
+}
+
+// updateResolutionUsageAndPointers figures out how many bytes are
+// referenced and unreferenced in the merged branch by this resolution
+// (if needed), and adds referenced and unreferenced pointers to a
+// final `resolutionOp` as necessary. It should be called before the
+// block changes are unembedded in md.  It returns the list of blocks
+// that can be remove from the flushing queue, if any.
+func (cr *ConflictResolver) updateResolutionUsageAndPointers(
+	ctx context.Context, lState *lockState, md *RootMetadata,
+	bps *blockPutState, unmergedChains, mergedChains *crChains,
+	mostRecentUnmergedMD, mostRecentMergedMD ImmutableRootMetadata,
+	isLocalSquash bool) (
+	blocksToDelete []kbfsblock.ID, err error) {
+
+	// Track the refs and unrefs in a set, to ensure no duplicates
+	refs := make(map[BlockPointer]bool)
+	unrefs := make(map[BlockPointer]bool)
+	for _, op := range md.data.Changes.Ops {
+		// Iterate in reverse since we may be deleting references as we go.
+		for i := len(op.Refs()) - 1; i >= 0; i-- {
+			ptr := op.Refs()[i]
+			// Don't add usage if it's an unembedded block change
+			// pointer.  Also, we shouldn't be referencing this
+			// anymore!
+			if unmergedChains.blockChangePointers[ptr] {
+				cr.log.CDebugf(ctx, "Ignoring block change ptr %v", ptr)
+				op.DelRefBlock(ptr)
+			} else {
+				refs[ptr] = true
+			}
+		}
+		// Iterate in reverse since we may be deleting unrefs as we go.
+		for i := len(op.Unrefs()) - 1; i >= 0; i-- {
+			ptr := op.Unrefs()[i]
+			unrefs[ptr] = true
+			delete(refs, ptr)
+			if _, isCreateOp := op.(*createOp); isCreateOp {
+				// The only way a create op should have unref blocks
+				// is if it was created during conflict resolution.
+				// In that case, we should move the unref to a final
+				// resolution op, so it doesn't confuse future
+				// resolutions.
+				op.DelUnrefBlock(ptr)
+				md.data.Changes.Ops =
+					addUnrefToFinalResOp(md.data.Changes.Ops, ptr)
+			}
+		}
+		for _, update := range op.allUpdates() {
+			if update.Unref != update.Ref {
+				unrefs[update.Unref] = true
+				delete(refs, update.Unref)
+				refs[update.Ref] = true
+			}
+		}
+	}
+
+	if isLocalSquash {
+		// Local squashes can just use the bytes and usage from the
+		// latest unmerged MD, and we can avoid all the block fetching
+		// done by `updateResolutionUsage()`.
+		md.SetDiskUsage(mostRecentUnmergedMD.DiskUsage())
```

Comment:
Might be clearer to have some temp variables here with shorter names, since the long names kind of run together:

```
unmergedUsage := mostRecentUnmergedMD.DiskUsage()
mergedUsage := mostRecentMergedMD.DiskUsage()
mergedMDUsage := mostRecentMergedMD.MDDiskUsage()

md.SetDiskUsage(unmergedMDUsage)
if unmergedUsage > mergedUsage {
  ...
}
...
```